existential-vs-universal-quantification/article.md

```diff
@@ -0,0 +1,334 @@
+
+```Haskell
+veryLongFunction :: a -> b -> c -> d -> e -> f -> g -> h -> i -> j
+veryLongFunction = ...
+
+func = veryLongFunction @_ @_ @_ @_ @_ @_ @_ @_ @_ @Integer ...
+```
+
+Quite long, right?
+However, this can be simplified with explicit `forall`:
+
+```Haskell
+veryLongFunction :: forall j a b c d e f g h i. a -> b -> c -> d -> e -> f -> g -> h -> i -> j
+veryLongFunction = ...
+
+func = veryLongFunction @Integer ...
+```
+
+Since `j` is the first type variable in the declaration of `veryLongFunction`, you can specify only it.
+This is useful not only in such big examples but also when your function requires any of its arguments to be specified on usage.
+
+### Supporting other extensions
+
+On its own, `ExplicitForAll` doesn't do a lot.
+
+But you frequently need it to enable other extensions. Some of them just won't work without the `forall` quantifier. :)
+
+The most common one is `ScopedTypeVariables`, which allows you to pass a type variable from an outer function to one defined in a where clause.
+
+The code below will not compile because the compiler thinks that the `a` in `example`'s signature and the `a` in `pair`'s are two different types.
+
+```haskell
+example :: a -> [a] -> [a]
+example x rest = pair ++ rest
+  where
+    pair :: [a]
+    pair = [x, x]
+```
+
+To fix it, you need to use the `ScopedTypeVariables` extension and also add a `forall` quantifier to the `example` function, since the extension won't work without it.
+
+```haskell
+example :: forall a. a -> [a] -> [a]
+```
+You might also notice how `forall` was useful with the `TypeApplication` extension.
+Other extensions that somehow benefit from the usage of `ExplicitForAll` are `LiberalTypeSynonyms`, `RankNTypes`, and many more.
+
+## Existential quantification
+
+As I said earlier, besides universally quantified variables, Haskell also supports existential types.
+They are also created with the `forall` keyword.
+
+They can be implemented this way because these two constructions – `(exists x. p x) -> q` and `forall x. (p x -> q)` – are equivalent in terms of first-order predicate logic.
+For a theoretical proof of this statement, you can check this [thread](https://stackoverflow.com/questions/10753073/whats-the-theoretical-basis-for-existential-types?rq=1).
+
+### Usage
+
+In order to proceed further, we'll have to enable the `RankNTypes` extension. This lets us write higher-rank types.
+
+<hr>
+
+**Higher-rank types**
+
+By default, all types are rank-1. In a rank-1 type, a `forall` cannot appear on the left of an arrow (keep in mind that the `->` operator has precedence over `.`)
+
+For example, this function has a rank-1 type:
+
+```haskell
+f :: forall a b. a -> b -> a
+
+-- Or, to make the precedence explicit:
+f :: forall a b. (a -> b -> a)
+```
+
+With the `RankNTypes` extension, we can write rank-2 types – types where `forall` appears to the left of one arrow:
+
+```haskell
+f :: (forall a. Show a => a -> IO ()) -> IO ()
+```
```

Comment:
I think it is much, much simpler to say that rank 2 types take as a parameter a rank 1 type, just like higher-order functions take as arguments regular functions. 

Would also show a rank 3 function to hammer it down.

Reply:
@PruStephan This is still unanswered.

Reply:
Some more good examples from this [StackOverflow answer](https://stackoverflow.com/a/22362474/857807):

```
Rank 0: Int
Rank 1: forall a. a -> Int
Rank 2: (forall a. a -> Int) -> Int
Rank 3: ((forall a. a -> Int) -> Int) -> Int
```

Which seem to match the examples in SPJ et al's paper ["Practical type inference for arbitrary-rank types"](https://www.microsoft.com/en-us/research/publication/practical-type-inference-for-arbitrary-rank-types/?from=http%3A%2F%2Fresearch.microsoft.com%2Fusers%2Fsimonpj%2Fpapers%2Fhigher-rank%2F):

![image](https://user-images.githubusercontent.com/718166/172882726-3fca6a09-51f7-420f-8935-7f28ba978d30.png)

So, instead of saying "By default, all types are rank-1", we should really be saying that, by default, Haskell supports rank-0 types (aka monomorphic types) and all polymorphic types are rank-1.

Reply:
Done here: https://github.com/serokell/blog-posts/pull/4/commits/d50ff4f65468c46dc454159dfef9d95813832ea2